### PR TITLE
WIP oCaml language support

### PR DESCRIPTION
**Pull request by dvberkel**
This is very much work in progress. Feedback is much appreciated.

### CODE REVIEW

**Comment by dvberkel** [comment]
Progress made towards #148.

I haven't tested this and I still need to require `klipse.lang.ocaml` in the places mentioned at [How to add a language to klipse](https://github.com/viebel/klipse/wiki/How-to-add-a-language-to-klipse). I figured that early feedback is most beneficial. Any tips, hints and feedback is much appreciated

**Comment by dvberkel** [comment]
How are individual languages tested?

**Comment by viebel** [comment]
Take a look at the test page for javascript: https://github.com/viebel/klipse/blob/master/resources/public/js-dbg.html

File: src/klipse/lang/ocaml.cljs [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+  (go
+    (try
+      ocaml-to-js
+      (catch :default o
+        (str o)))))
+
+(def eval-opts {:editor-in-mode "text/ocaml"
+                :editor-out-mode "javascript"
+                :eval-fn eval-ocaml
+                :external-csripts [(codemirror-mode-src "ocaml") "http://bloomberg.github.io/bucklescript/js-demo/exports.js"]
```
Comment (dvberkel) [inline src/klipse/lang/ocaml.cljs]: Sure! Will do as soon as possible

Comment (dvberkel) [inline src/klipse/lang/ocaml.cljs]: It landed in pull request #156

**Comment by viebel** [comment]
I have made a couple of improvements and stuff is starting to work.
You can cherry pick this commit https://github.com/viebel/klipse/commit/0f23d415b2495f59ceadc38865a5f14d5dee295d

**Comment by dvberkel** [comment]
What needs to be done before this can be merged. I have some time later tonight to work on this, but have no clear direction to work in

**Comment by viebel** [comment]
So cool that you have time tonight for ocaml.

For the moment, the transpilation works fine.

The evaluation works in a weird way. As far as I understand (and I have no knowledge at all in Ocaml) in order for a snippet to be evaluated properly, it must end by a `let () = ...` statement and this statement must have a side effect e.g. `JS.log`.

The trick I did, in order to see the `Js.log` result is with `my-with-redefs` (redirect console.log to the result snippet).

Also, it seems that symbols are not shared between snippets.

The best would be for you to write a blog post that teaches the basics of ocaml and see how convenient it is to use the klipse snippet. And if something is not convenient, then fix the ocaml support in klipse.

Makes sense?

**Comment by dvberkel** [comment]
Totally makes sense, although I am confused. Did oCaml support already land in the master branch? Or should I build a local copy of the plugin?

**Comment by viebel** [comment]
Not yet on master.
You need to build a local copy with `./scripts/build`

**Comment by viebel** [comment]
Any progress @dvberkel ?

I have played with it a bit a found a couple of issues:

1. In order to display something, one has to use `Js.log` or end the code snippet with something like `let a=1+1;`. It would be better to evaluate statements like `1+1`.

2. Variables definitions are not shared between code snippets.

3. The types of the data is not displayed.

**Comment by viebel** [comment]
I have opened an issue on tryocaml https://github.com/OCamlPro/tryocaml/issues/167
They might help

**Comment by dvberkel** [comment]
I made some progress, but got side tracked integrating klipse in my blog. I wrote a [blog post](http://fifth-postulate.nl/blog/2016-12-10.4-Number-Game.html) with clojure in mind to see what is to be expected.

I will take a look into tryocaml and their environment

**Comment by viebel** [comment]
@dvberkel Thanks for your initiative. Now Klipse supports Ocaml and ReasonML. See http://blog.klipse.tech/ocaml/2017/10/05/blog-ocaml.html

**Comment by dvberkel** [comment]
Congratulations! Sorry that I couldn't find the time to bring this to fruitation

**Pull request closed without merging**